File: Placar09/file_recover_test.py
import math
def longestPrefixSuffix(s) : 
    n = len(s) 
      
    for res in range(n // 2, 0, -1) : 
          
        prefix = s[0: res] 
        suffix = s[n - res: n] 
          
        if (prefix == suffix) : 
            return res 
              
    return 0

while (1):
    row = input().split(" ")
    length = int(row[0])
    string = row[1]
    lenString = len(string)

    if length == -1: break
    
    if lenString > length:
        print(str(0))
    else:
        prefixLen = longestPrefixSuffix(string)

        tst = math.floor((length - prefixLen) / (lenString - prefixLen))
        print(tst)
        

File: Placar09/file_recover_test.cpp
#include <bits/stdc++.h>
using namespace std;

int longestPrefixSuffix(string s) 
{ 
    int n = s.length(); 
  
    int lps[n]; 
    lps[0] = 0; // lps[0] is always 0 
  
    // length of the previous 
    // longest prefix suffix 
    int len = 0; 
  
    // the loop calculates lps[i] 
    // for i = 1 to n-1 
    int i = 1; 
    while (i < n) 
    { 
        if (s[i] == s[len]) 
        { 
            len++; 
            lps[i] = len; 
            i++; 
        } 
        else // (pat[i] != pat[len]) 
        { 
            // This is tricky. Consider 
            // the example. AAACAAAA 
            // and i = 7. The idea is 
            // similar to search step. 
            if (len != 0) 
            { 
                len = lps[len-1]; 
  
                // Also, note that we do 
                // not increment i here 
            } 
            else // if (len == 0) 
            { 
                lps[i] = 0; 
                i++; 
            } 
        } 
    } 
  
    int res = lps[n-1]; 
  
    // Since we are looking for 
    // non overlapping parts. 
    return res;
} 

int main() {
    while(true) {
        int length, prefixLen;
        string str;
        cin >> length >> str;

        if(length == -1) break;
        
        if(length < str.size()) {
            cout << 0 << endl;
        } else {
            prefixLen = longestPrefixSuffix(str);
            cout << (length - prefixLen) / (str.size() - prefixLen) << endl;
        }

    }
    return 0;
}
